tpp-app/src/services/cycle/CycleService.js

```diff
@@ -0,0 +1,290 @@
+    var current = searchFrom;
+    var yesterday = subDays(current, 1);
+    var twoDaysEarlier = subDays(current, 2);
+
+    if (!calendar){
+      calendar = await getCalendarByYear(current.getFullYear());
+    }
+
+    let dateSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth()+1, current.getFullYear());
+    let yesterdaySymptoms = new Symptoms();
+    let twoDaysEarlierSymptoms = new Symptoms()
+
+    var noFlowToday = (dateSymptoms.flow === FLOW_LEVEL.NONE || dateSymptoms.flow === null) ;
+    var noFlowYesterday = (yesterdaySymptoms.flow === FLOW_LEVEL.NONE || yesterdaySymptoms.flow === null) ;
+    var flowTwoDaysEarlier = (twoDaysEarlierSymptoms.flow !== null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE)
+
+
+    while(!(noFlowToday && noFlowYesterday && flowTwoDaysEarlier)){
+
+      var tomorrow = addDays(current, 1);
+      twoDaysEarlier = yesterday;
+      yesterday = current;
+      current = tomorrow;
+
+
+      twoDaysEarlierSymptoms = yesterdaySymptoms;
+      yesterdaySymptoms = dateSymptoms;
+      dateSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth() + 1, current.getFullYear());
+      noFlowToday = (dateSymptoms.flow === FLOW_LEVEL.NONE || dateSymptoms.flow === null) ;
+      noFlowYesterday = (yesterdaySymptoms.flow === FLOW_LEVEL.NONE || yesterdaySymptoms.flow === null) ;
+      flowTwoDaysEarlier = (twoDaysEarlierSymptoms.flow !== null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE)
+    }
+
+
+
+    //return twoDaysEarlier since pattern searching for is _ _ X where X is the period
+    return twoDaysEarlier;
+
+
+}
+export {getNextPeriodEnd};
+
+/**
+ * Gets the most recent period start date for a given date (searchFrom)
+ * @param {Date} searchFrom Date from which to find the last period start
+ * @param {Object} calendar The object containing the symptoms for this year, last year, and next year. Optional.
+ * @return {Promise} Resolves into Date that is the most recent day that a period started, relative to searchFrom
+ */
+async function getLastPeriodStart(searchFrom, calendar = null){
+    var current = searchFrom;
+    var tomorrow = addDays(current, 1);
+    var twoDaysLater = addDays(current, 2);
+    if (!calendar){
+      calendar = await getCalendarByYear(current.getFullYear());
+    }
+
+    let dateSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth()+1, current.getFullYear());
+    let tomorrowSymptoms = new Symptoms();
+    let twoDaysLaterSymptoms = new Symptoms()
+
+    var noFlowToday = (dateSymptoms.flow === FLOW_LEVEL.NONE || dateSymptoms.flow === null) ;
+    var noFlowTomorrow = (tomorrowSymptoms.flow === FLOW_LEVEL.NONE || tomorrowSymptoms.flow === null) ;
+    var flowTwoDaysLater = (twoDaysLaterSymptoms.flow !== null && twoDaysLaterSymptoms.flow !== FLOW_LEVEL.NONE)
+
+
+    while(!(noFlowToday && noFlowTomorrow && flowTwoDaysLater)){
```

Comment:
Correct me if I'm wrong but I think this while loop may not exit in some cases and be stuck in infinite loop. If we have a new user who has no period logged, this could be searching past dates forever for the _ _ X pattern, which does not exist. 

Also, I am not convinced subDays() is the best way of 'taking a step backwards' when iterating since you are always taking a step of one day backwards. A better way would be to take larger steps backwards by not checking dates where we know the pattern  _ _ X would not match. This strategy would be to identify dates where there are periods so we are guaranteed to match the "X" part of the pattern. 

What you have: 
- _ _ X is a pattern that walks backwards in time one day at a time 
- The strategy is not wrong, but there is a lot of unnecessary comparisons  
- If period dates in 2022 are [Jan 10, Jan 11, Jan 12, Jan 13], and searchFrom=Feb 28, you are making ~40 comparisons 

What I am suggesting: 
- Find list of period dates in current year using a helper. If year=2022 retrieve list of all period dates in 2022 ex: [Jan 10, Jan 11, Jan 12, Jan 13]
- searchFrom = Feb 28 
- _ _ X is a pattern that walks backwards in time by matching the X in the pattern to the list of period dates
  - check flowTwoDaysLater=Jan 13's flow, noFlowTomorrow=Jan 12's flow, noFlowToday=Jan 11's flow. It's not a match to _ _ X pattern.  Set the "X" in the pattern to the next date, Jan 12 to check Jan 10, Jan 11, Jan 12 if it matches _ _ X. 
  - And so on...until you get a match of _ _ X to Jan 8, Jan 9, Jan 10
  - This way instead of taking a step of one day backwards, we optimize by taking larger steps backwards to skip days where we know the pattern won't match giving us more of a chance of success in matching the desired pattern 
  - Instead of making ~40 comparisons, you only made 4 comparisons 

I think this would also solve the problem of the infinite while loop if there is no last period start since you can add a condition to the while loop to exit when you run out of period dates in the current year. 

We can chat more, this is big infodump 💩

Reply:
Same logic can apply to ```getNextPeriodEnd()``` when searching for X _ _ pattern match

Reply:
so have a function (let's say `getPeriodDates`) that returns an array of Date objects that contains all the dates that there are periods in sequential order right? And then you only need to check for the pattern on these period dates returned from the function? 

Isn't that the same amount of comparisons at the end though? If we call the number of dates in a year _n_, then `getPeriodDates` has to do _n_ accesses to guarantee it has all the period days, and then do some more comparisons in `getLastPeriodStart` (O(n)). 

Also yeah you are totally right this does cause an infinite loop if the user has no period logged. I think swapping to ur idea would deal with the issue cleanly

Reply:
Actually I think it does mean markedly less computations for some functions since we can call `getPeriodDates` once and pass it in to the relevant functions in the cases where we call them multiple times (Basically just `GETCycleHistoryByYear`). 

The one-off functions like `GETPeriodDate` will be a bit slower if u don't give them a pre-computed list, but I think that's fine.

Reply:
Ok I refactored `GETPeriodDay` and `GETMostRecentPeriodStartDate` to use an array of periods for the year.

I also refactored `GETCycleHistoryByYear` to use an array of periods. I still use the step-by-step iteration approach for the very first and very last period of the year, so that it will be able to read periods like "Dec 27 2021" -> "Jan 2cnd 2022" as one interval in both `GETCycleHistoryByYear(2021)` and `GETCycleHistoryByYear(2022)`. I changed docstring of these step-by-step methods to be explicit that the `Date` to search from should be a period day, and wrote the code so they can't get called in a calendar that hasn't logged any periods.